explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,260 @@
+
+```
+-fauto-profile=/path/to/openssl-nonpgo.gcov
+```
+
+Make sure to adjust the path to the `openssl-nonpgo.gcov` file.
+
+Also note that enabling PGO will make GCC automatically enable several optimization passes that are usually disabled, even when using `-O2`.  This can lead to new warnings/errors in the code, which might require either the code to be adjusted or the use of extra compilation flags to suppress these warnings.
+
+After the software is built, you can install this new version and run the benchmark tests again to compare results.
+
+### Our results
+
+At first sight it might appear that our results don't have much to show for themselves, but they are actually very interesting and allow us to discuss some more things about PGO.
+
+First, it is not a silver bullet.  It will not work for every scenario, and there might be cases when it is not justified to use it.  For example, if you are dealing with a very generic program that doesn't have a single representative workload.  PGO is great when applied to specialized cases, but may negatively impact the performance of programs that can be used in a variety of ways.
+
+It is also really important to be rigorous when doing the profiling of the program you want to optimize.  Here are a few tips that might help obtain more confident results:
+
+* Make sure to perform multiple runs of the same workload.  This helps to eliminate possible outliers as well as decrease the contribution of external factors (like memory and cache times).  In our example, we performed 5 consecutive runs without PGO and 5 consecutive runs with PGO.  Had we performed more runs, some of the outliers we see in the results would be even less noticeable.
+
+* Try to configure your system to have as little load as possible.  If you have other programs running in the background, they will certainly affect the final measurements in non-deterministic ways.
+
+* Another important thing to consider is whether you are dealing with code that has already been highly optimized.  In fact, this is exactly the case with our example: OpenSSL is an ubiquitous cryptographic library that has received *a lot* of attention from several developers, and it is expected that cryptographic libraries are extremely optimized.  This is the major factor that explains why our results were not so great: the library is already pretty performant as it is!
+
+### Conclusion
+
+PGO is an interesting and complex topic that certainly catches the attention of experienced software developers looking to extract every bit of performance gain.  It can certainly help optimize programs, especially those that perform specialized work.  It is not a one-size-fits-all solution, though, and as with every complex technology, its use needs to be studied and justified, and its impact on the final application should be monitored and analyzed.
+
+## More information
+
+LINK TO BLOG POST HERE
```

Comment:
Don't forget to link :D

Reply:
Thanks!  Since the blog post will be available later, I have temporarily removed this subsection and will reintroduce it when I have the final link.